### PR TITLE
Added initial endpoints

### PR DESCRIPTION
**Pull request by varvarenique**
https://trello.com/c/IDEtbh6y/2-create-a-simple-flask-application-with-several-endpoints

### CODE REVIEW

**Comment by minotru** [comment]
Please think about a proper name for the PR and, in future, git branch. The name should make it clear what is going on inside. For example, PR name can be "Added initial endpoints", and it can be somehow reflected in branch name too (e.g. `initial-endpoints`).

Another good practice is to link the PR with the task it relates to, e.g. paste Trello task link in PR description, otherwise it is not clear what was the scope of PR, what was accomplished.

To summarize:
1. Sensible PR name and, for the future, branch name
2. Sensible PR description or, at least, link to the task

File: app.py [hunk 1]
```diff
@@ -1,2 +1,57 @@
...
+with open("restaurants.json", "r") as read_file:
+    data = json.load(read_file)
+
+restaurants = data['restaurants']
+
+
+@app.route('/api/restaurants', methods=['GET', 'POST'])
+def all_restaurants():
+    if request.method == 'POST':
+        content = request.get_json(silent=True)
```
Comment (minotru) [inline app.py]: `silent=True` forces Flask to silently return `None` even if an error occurred (JSON was absent or malformed). Is it what you wanted to achieve? POST body is required here.

In general, more common usage is: `content = request.json`. - it will automatically parse body as JSON and raise ValidationError if JSON is missing or malformated

File: app.py [hunk 2]
```diff
@@ -1,2 +1,57 @@
...
+        print(type(content))
```
Comment (minotru) [inline app.py]: Avoid any `print` statements unless it is a console application where you do need to print something to console. You can have them for debug purposes, during development, but the code in PR should be clean

File: app.py [hunk 3]
```diff
@@ -1,2 +1,57 @@
...
+        if content is not None:
+            print(content)
+            restaurants.append(content['restaurant_name'])
+        response = app.response_class(
+            response=json.dumps({'restaurants': restaurants}),
+            status=201,
+            mimetype='application/json'
+        )
+        return response
```
Comment (minotru) [inline app.py]: Flask has a nice pattern when you want to return json response:
```python
from flask import jsonify

@app.route("/api/test")
def route():
   ....
   return jsonify(my_object), status_code # e.g. jsonify({'restaurants': restaurants}), 201
```

File: app.py [hunk 4]
```diff
@@ -1,2 +1,57 @@
...
+    else:
```
Comment (minotru) [inline app.py]: Imagine that you are reading this code for the first time. you see handler like:
```python
if request.method == 'POST':
   ...
else:
   ...
```
so you think: "well, what does this `else` mean?", you look above and see `methods=['GET', 'POST']` "ah, hm, this  endpoint supports 2 operations, first one was `POST`, then another is `GET`"

then you added one more HTTP method: `methods=['GET', 'POST', 'DELETE']`
and there is:
```python
if request.method == 'POST':
   ...
elif request.method == 'DELETE':
   ...
else:
  ...
```

I guess you see what I mean.
So it is better to be explicit and write like `elif request.method == 'GET'`

OR you can use Flask's [MethodView ](https://flask.palletsprojects.com/en/1.1.x/views/#method-based-dispatching)and then there is no need in these `if-else` at all :)
(to my mind, it makes sense to use them if there are several HTTP methods for the same route)

File: restaurants.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (minotru) [inline restaurants.json:1]: could you please add more fields to restaurant? in real-world, the restaurant has an address, working hours... - as you described your entities in requirements document.
also, any entity should have its ID, so that we can index it and find it

**Review by minotru — CHANGES_REQUESTED** [review]
please create `requirements.txt`

File: app.py (hunk 1 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 2 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 3 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 4 above)
Comment (varvarenique) [reply app.py]: fixed

File: restaurants.json (hunk 5 above)
Comment (varvarenique) [reply restaurants.json:1]: fixed

File: restaurants.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  "restaurants":{
```
Comment (minotru) [inline restaurants.json]: it is more common to represent such data as a `list` not a `dictionary`. every JSON record should be self-contained, i.e. it is clear what means what without extra context.

 imagine I gave you data like `[12, 13, 28]` - do you have an idea about what these numbers mean?
instead, it can be `{"month": 12, "numberOfCandies": 13, "age": 28}`


so the common way to write such data is like
```json
{
  "restaurants": [
    {
      "id": 0,
      "name": "Vasilki",
      "location": "Minsk"
       ....
    }
  ]
}
```

please use a format like this to store the data in JSON and in Python

File: app.py [hunk 7]
```diff
@@ -1,2 +1,40 @@
...
+restaurants = data['restaurants']
+
+
+class RestaurantEndpoint(MethodView):
+    def get(self):
+        return jsonify({'restaurants': restaurants}), 200
+
+    def post(self):
+        content = request.json
+        if content is not None:
```
Comment (minotru) [inline app.py]: the point of `POST` method is to create a new record. if the body is empty (`if content is None`), then it just does not make sense, it is wrong input. so it would be more logical to raise something like Flask's `ValidationError` if body is empty

File: app.py [hunk 8]
```diff
@@ -1,2 +1,40 @@
...
+app = Flask(__name__)
+with open("restaurants.json", "r") as read_file:
+    data = json.load(read_file)
+
+restaurants = data['restaurants']
+
+
+class RestaurantEndpoint(MethodView):
+    def get(self):
+        return jsonify({'restaurants': restaurants}), 200
```
Comment (minotru) [inline app.py]: `Restaraunt` endpoint is responsible for the resource `restaurants`.
so imagine you see an endpoint like `GET /api/pets`. `pets` sounds like a `list of pet`.
in this case, it is better to respond with a list rather than a dict that has `restaurants` property that is a list of `restaraunt`

File: app.py [hunk 9]
```diff
@@ -1,2 +1,40 @@
...
+
+class RestaurantItemEndpoint(MethodView):
+    def get(self, restaurant_id):
+        return jsonify({'info': restaurants[str(restaurant_id)]}), 200
+
+    def put(self, restaurant_id):
+        content = request.json
+        if content is not None:
+            restaurants[str(restaurant_id)] = content['restaurant_name']
+        return jsonify({'restaurants': restaurants}), 200
```
Comment (minotru) [inline app.py]: `RestaurantItemEndpoint` processes single `restaraunt` info, so it is better to respond with a single `restaurant` that was changed, not a list of all restaurants.

when you think about proper input-output, take a look at your URLs:
* `/api/restaurants` -  restaurants is plural, restaurants is some collection (list), so operations on `/api/restaurants` level affect all collection
* `/api/restaurants/<int:restaurant_id>`- this endpoint refers to a particular restaraunt, so it should get/update/delete only one item from collection

**Review by minotru — CHANGES_REQUESTED** [review]

File: restaurants.json (hunk 6 above)
Comment (varvarenique) [reply restaurants.json]: fixed

File: app.py (hunk 7 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 8 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 9 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py [hunk 10]
```diff
@@ -1,2 +1,46 @@
-from flask import Flask, request
+from flask import Flask, request, jsonify
 import json
```
Comment (minotru) [inline app.py:1]: small comment: 
it is a good practice to order the inputs in this way:
1. standard library packages
2. installed packages
3. your own packages

e.g.
```python
import json

from flask import Flask, request, jsonify

import mymodule
```

File: restaurants.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      "name": "Vasilki",
+      "address": "Minsk",
+      "work time": "Monday-Sunday: 08:00 - 23:45",
+      "phone number": "+375297777777"
+    },
+    {
+      "id": 1,
+      "name": "Mama Doma",
+      "address": "Minsk",
+      "work time": "Monday-Sunday: 10:00 - 22:00",
```
Comment (minotru) [inline restaurants.json]: it is better to avoid spaces in names, e.g. use `workTime`, `work_time`

File: app.py [hunk 12]
```diff
@@ -1,2 +1,46 @@
...
+
+
+class RestaurantItemEndpoint(MethodView):
+    def get(self, restaurant_id):
+        return jsonify(restaurants[restaurant_id]), 200
+
+    def put(self, restaurant_id):
+        content = request.json
+        answer = 200
+        if content is not None:
```
Comment (minotru) [inline app.py]: `put` is to update some item, if body is `None` then it is a malformed request.
it is like:
```
- hey, update this 
- update what?
```

File: app.py [hunk 13]
```diff
@@ -1,2 +1,46 @@
...
+            if len(restaurants) <= restaurant_id:
+                answer = 201
```
Comment (minotru) [inline app.py]: `PUT` updates some existing item. so PUT is "I want to change this restaurant info, e.g. change address". and `POST` is to create an item

File: app.py [hunk 14]
```diff
@@ -1,2 +1,46 @@
...
+                restaurants.append(content)
+            else:
+                restaurants[restaurant_id] = content
+        return jsonify(restaurants[restaurant_id]), answer
```
Comment (minotru) [inline app.py]: `restaraunt_id` does not have to follow order of items in the list, `restaraunt_id` is `item.id`.
 you should scan the list and find an item where `item.id == restaurant_id`

**Review by minotru — CHANGES_REQUESTED** [review]

File: app.py (hunk 10 above)
Comment (varvarenique) [reply app.py:1]: fixed

File: restaurants.json (hunk 11 above)
Comment (varvarenique) [reply restaurants.json]: fixed

File: app.py (hunk 12 above)
Comment (varvarenique) [reply app.py]: will fix in another PR

File: app.py (hunk 13 above)
Comment (varvarenique) [reply app.py]: fixed

File: app.py (hunk 14 above)
Comment (varvarenique) [reply app.py]: fixed
